### PR TITLE
ZBUG-4531 : Skip Imap UID range search

### PR DESCRIPTION
**Pull request by chitranshu198**
[ZBUG-4531 ](https://synacor.atlassian.net/browse/ZBUG-4531)
The IMAP search query which includes a UID range (e.g., UID 1:x,y:*), results in the generation of long SQL queries and causing longer execution time for the search.

  Solution : Check if Sequence search's sub sequence contains UID range pattern; if it does, skip processing.
  Added a LC config to make this check configurable.

### CODE REVIEW

**Review by preshitaagrawal — APPROVED** [review]

**Review by rcyarrapothu — APPROVED** [review]

**Review by shrutig0510 — APPROVED** [review]

File: store/src/java/com/zimbra/cs/imap/ImapSearch.java [hunk 1]
```diff
@@ -33,6 +36,8 @@ abstract class ImapSearch {
     protected abstract String toZimbraSearch(ImapFolder i4folder) throws ImapParseException;
     protected abstract ImapMessageSet evaluate(ImapFolder i4folder) throws ImapParseException;
     protected boolean requiresMODSEQ()  { return false; }
+    private static final String UID_SEARCH_PATTERN = "^1:(\\d+),(\\d+):\\*$";
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/imap/ImapSearch.java:40]: please add a comment for UID pattern that we going to ignore in the search

File: store/src/java/com/zimbra/cs/imap/ImapSearch.java [hunk 2]
```diff
@@ -167,7 +172,16 @@ static class AndOperation extends LogicalOperation {
         @Override
         protected String toZimbraSearch(ImapFolder i4folder) throws ImapParseException {
             StringBuilder search = new StringBuilder("(");
+            Pattern compiledUidPattern = Pattern.compile(UID_SEARCH_PATTERN);
             for (ImapSearch i4search : mChildren) {
+                if (i4search instanceof SequenceSearch
+                        && isUidRangeSearchIgnored
+                        && null != ((SequenceSearch) i4search).mSubSequence) {
+                    Matcher uidMatcher = compiledUidPattern.matcher(((SequenceSearch) i4search).mSubSequence);
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/imap/ImapSearch.java]: there is no use of `uidMatcher` further can combine this with if check below `compiledUidPattern.matcher(((SequenceSearch) i4search).mSubSequence).find()`

Comment (chitranshu198) [reply store/src/java/com/zimbra/cs/imap/ImapSearch.java]: Fixed.

File: store/src/java/com/zimbra/cs/imap/ImapSearch.java (hunk 1 above)
Comment (chitranshu198) [reply store/src/java/com/zimbra/cs/imap/ImapSearch.java:40]: Done. Added details of the pattern along with example

**Review by Gopal-Moolchandani — APPROVED** [review]

File: store/src/java/com/zimbra/cs/imap/ImapSearch.java [hunk 3]
```diff
@@ -167,7 +173,15 @@ static class AndOperation extends LogicalOperation {
         @Override
         protected String toZimbraSearch(ImapFolder i4folder) throws ImapParseException {
             StringBuilder search = new StringBuilder("(");
+            Pattern compiledUidPattern = Pattern.compile(UID_SEARCH_PATTERN);
             for (ImapSearch i4search : mChildren) {
+                if (i4search instanceof SequenceSearch
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/imap/ImapSearch.java:178]: we can combine both these if blocks

**Review by dasiyogesh — APPROVED** [review]

File: store/src/java/com/zimbra/cs/imap/ImapSearch.java (hunk 3 above)
Comment (chitranshu198) [reply store/src/java/com/zimbra/cs/imap/ImapSearch.java:178]: Fixed

**Pull request merged**